### PR TITLE
*: optimize DDL history http API, reduce memory

### LINKED ISSUE #35838: server: reduce the memory usage used by `getAllHistoryDDL`
## Enhancement

TiDB provided a restful API to query all DDL jobs in the history: `http://{TiDBIP}:10080/ddl/history`.

The initial intention of the API is to debug TiDB in an easy way. However, it has been used by some old ecosystem tools like ['drainer'](https://github.com/pingcap/tidb-binlog).

When there are a lot of DDL jobs for a long-running cluster, DDL jobs list will be long and it takes a large amount of memory when querying `http://{TiDBIP}:10080/ddl/history`.

For the compatibility with old tools, it is reasonable to try reducing the memory usage of `getAllHistoryDDL`.

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #35838

Problem Summary:

### What is changed and how it works?

We can't cut down the memory cost for `GetAllDDLHistory`, but we can do better for 

http://127.0.0.1:10080/ddl/history?limit=N

The current implementation for limit N is, get all the data, sort them, and take n items.

I also add a paging like API, now we can use `start_job_id` together with `limit`

http://127.0.0.1:10080/ddl/history?start_job_id=xxx&limit=yyy

to avoid using the GetALL API the fetch the whole DDL history data.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

curl http://127.0.0.1:10080/ddl/history?start_job_id=32&limit=5

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Optimize DDL history HTTP API, and add support for 'start_job_id' parameter.
```

### CODE REVIEW

File: server/http_handler.go [hunk 1]
```diff
@@ -1251,50 +1252,51 @@ func (h tableHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		jobID, err = strconv.Atoi(jobValue)
 		if err != nil {
 			writeError(w, err)
 			return
 		}
-
-		if lid < 1 {
-			writeError(w, errors.New("ddl history limit must be greater than 1"))
+		if jobID < 1 {
+			writeError(w, errors.New("ddl history start_job_id must be greater than 1"))
```
Comment (tangenta) [inline server/http_handler.go:1264]:
```suggestion
			writeError(w, errors.New("ddl history start_job_id must be greater than 0"))
```

File: server/http_handler.go [hunk 2]
```diff
@@ -1251,50 +1252,51 @@ func (h tableHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+			writeError(w, err)
 			return
 		}
-
-		jobsLen := len(jobs)
-		if jobsLen > lid {
-			start := jobsLen - lid
-			jobs = jobs[start:]
+		if limitID < 1 {
+			writeError(w, errors.New("ddl history limit must be greater than 1"))
```
Comment (tangenta) [inline server/http_handler.go:1275]:
```suggestion
			writeError(w, errors.New("ddl history limit must be greater than 0"))
```

**Comment by tiancaiamao** [comment]
cat test1.go

```
package main

import (
	"fmt"
)

func main() {
	for i:=0; i<3000; i++ {
		fmt.Printf("create table t%d (id int);\n", i)
	}
	for i:=0; i<3000; i++ {
		fmt.Printf("alter table t%d add column (b int);\n", i)
	}
	for i:=0; i<3000; i++ {
		fmt.Printf("alter table t%d add index i_b(b);\n", i)
	}
	for i:=0; i<3000; i++ {
		fmt.Printf("alter table t%d drop index i_b;\n", i)
	}
	for i:=0; i<3000; i++ {
		fmt.Printf("drop table t%d;\n", i)
	}
}
```

go run test1.go > sql.txt
mysql -h 127.0.0.1 -u root -P 4000 -A test < sql.txt


cat test2.go
```
package main

import (
	"fmt"
	"io"
	"sync/atomic"
	"time"
	"net/http"
)

var QPS uint64

func main() {
	for i:=0; i<20; i++ {
		go worker()
	}

	
	ticker := time.NewTicker(time.Second)
	for {
		select {
		case <-ticker.C:
			v := atomic.LoadUint64(&QPS)
			fmt.Println("QPS =", v)
			atomic.StoreUint64(&QPS, 0)
		}
	}
}

func worker() {
	for ;; {
		resp, err := http.Get("http://127.0.0.1:10080/ddl/history?limit=20")
		if err != nil {
			panic(err)
		}
		io.Copy(io.Discard, resp.Body)
		resp.Body.Close()
		atomic.AddUint64(&QPS, 1)
	}
}
```

go run test2.go

Before the change:

```
QPS = 12
QPS = 19
QPS = 20
QPS = 20
QPS = 9
QPS = 14
QPS = 18
QPS = 19
QPS = 19
QPS = 14
QPS = 12
QPS = 16
QPS = 19
QPS = 19
QPS = 17
QPS = 15
QPS = 13
QPS = 17
QPS = 18
QPS = 19
QPS = 15
QPS = 16
QPS = 13
QPS = 17
QPS = 19
QPS = 17
QPS = 15
QPS = 14
QPS = 17
QPS = 17
QPS = 17
QPS = 18
QPS = 12
QPS = 18
QPS = 16
QPS = 17
QPS = 17
QPS = 17
QPS = 14
QPS = 16
QPS = 17
QPS = 16
QPS = 19
QPS = 16
QPS = 15
QPS = 16
QPS = 16
QPS = 18
QPS = 15
QPS = 15
```

After:

```
QPS = 4784
QPS = 4873
QPS = 4804
QPS = 4988
QPS = 4857
QPS = 4865
QPS = 4875
QPS = 4874
QPS = 4812
QPS = 4080
QPS = 4449
QPS = 4700
QPS = 4572
QPS = 4449
QPS = 4510
QPS = 4881
QPS = 4873
QPS = 4904
QPS = 4929
QPS = 5067
QPS = 4870
QPS = 4980
QPS = 4940
QPS = 4919
QPS = 5014
QPS = 4986
QPS = 4987
QPS = 4946
QPS = 5009
QPS = 5024
QPS = 4990
QPS = 5000
QPS = 4936
QPS = 5061
QPS = 4930
QPS = 5087
QPS = 4926
QPS = 4883
QPS = 4949
QPS = 4779
QPS = 4966
QPS = 4925
QPS = 5000
QPS = 5013
QPS = 4859
QPS = 5040
QPS = 5060
QPS = 4993
QPS = 5028
QPS = 4885
QPS = 5042
```

**Review by tangenta — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Comment by tiancaiamao** [comment]
Construct more ddl history data

```
mysql> admin show ddl jobs;
+--------+---------+------------+------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+--------+
| JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE   | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | CREATE_TIME         | START_TIME          | END_TIME            | STATE  |
+--------+---------+------------+------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+--------+
| 185770 | test6   | t857       | drop index | none         |     17866 |   119026 |         0 | 2022-08-04 22:12:59 | 2022-08-04 22:14:14 | 2022-08-04 22:14:14 | synced |
| 185769 | test3   | t914       | drop index | none         |      9129 |   119284 |         0 | 2022-08-04 22:12:58 | 2022-08-04 22:14:14 | 2022-08-04 22:14:14 | synced |
| 185768 | test    | t1167      | drop index | none         |         2 |   120366 |         0 | 2022-08-04 22:12:58 | 2022-08-04 22:14:14 | 2022-08-04 22:14:14 | synced |
| 185767 | test4   | t881       | drop index | none         |     13909 |   119072 |         0 | 2022-08-04 22:12:58 | 2022-08-04 22:14:14 | 2022-08-04 22:14:14 | synced |
| 185766 | test1   | t960       | drop index | none         |      6211 |   119484 |         0 | 2022-08-04 22:12:58 | 2022-08-04 22:14:14 | 2022-08-04 22:14:14 | synced |
| 185765 | test5   | t904       | drop index | none         |     17428 |   119486 |         0 | 2022-08-04 22:12:58 | 2022-08-04 22:12:58 | 2022-08-04 22:14:14 | synced |
| 185764 | test6   | t856       | drop index | none         |     17866 |   119014 |         0 | 2022-08-04 22:12:58 | 2022-08-04 22:12:58 | 2022-08-04 22:12:58 | synced |
| 185763 | test3   | t913       | drop index | none         |      9129 |   119272 |         0 | 2022-08-04 22:12:57 | 2022-08-04 22:12:58 | 2022-08-04 22:12:58 | synced |
| 185762 | test    | t1166      | drop index | none         |         2 |   120354 |         0 | 2022-08-04 22:12:57 | 2022-08-04 22:12:58 | 2022-08-04 22:12:58 | synced |
| 185761 | test4   | t880       | drop index | none         |     13909 |   119060 |         0 | 2022-08-04 22:12:57 | 2022-08-04 22:12:58 | 2022-08-04 22:12:58 | synced |
+--------+---------+------------+------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------------------+--------+
10 rows in set (0.02 sec)
```
The all dumped history data is 431M

Before this fix,  `curl 'http://127.0.0.1:10080/ddl/history?limit=20' > /tmp/xxx` takes 1.1G memory
And now it takes ... oh, the query execute so damn fast and I can't notice the memory change

**Pull request merged**